File: calculations.js
import { MONTHS_2026 } from "./data-schema.js";

const MONEY_EPSILON = 0.005;

export function getTransactionMonth(dateString) {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(String(dateString || ""))) {
    return "";
  }

  const date = new Date(`${dateString}T00:00:00`);
  if (Number.isNaN(date.getTime())) {
    return "";
  }

  const [year, month, day] = dateString.split("-").map(Number);
  if (date.getUTCFullYear() !== year || date.getUTCMonth() + 1 !== month || date.getUTCDate() !== day) {
    return "";
  }

  return `${year}-${String(month).padStart(2, "0")}`;
}

export function summarizeMonth(state, month) {
  const transactions = transactionsForMonth(state, month);
  const expenses = sum(transactions.filter((item) => item.type === "Expense").map((item) => item.amount));
  const income = sum(transactions.filter((item) => item.type === "Income").map((item) => item.amount));
  const categoryRows = summarizeCategoriesForMonth(state, month);
  const topCategory = categoryRows.slice().sort((a, b) => b.spend - a.spend)[0]?.category || "None";

  return {
    expenses,
    income,
    netCashflow: income - expenses,
    transactionCount: transactions.length,
    reimbursableTotal: sum(transactions.filter((item) => item.reimbursable).map((item) => item.amount)),
    essentialSpend: sum(transactions.filter((item) => item.essential && item.type === "Expense").map((item) => item.amount)),
    topCategory
  };
}

export function summarizeCategoriesForMonth(state, month) {
  return state.categories.map((category) => {
    const spend = sum(
      transactionsForMonth(state, month)
        .filter((transaction) => transaction.type === "Expense" && transaction.category === category.name)
        .map((transaction) => transaction.amount)
    );

    return {
      category: category.name,
      spend,
      budget: Number(category.defaultBudget) || 0
    };
  });
}

export function calculateAvailableToAssign(state, month) {
  const income = sum(transactionsForMonth(state, month).filter((item) => item.type === "Income").map((item) => item.amount));
  const funding = sum(state.monthlySetup.filter((item) => item.month === month).map((item) => item.monthlyTarget));
  return income - funding;
}

export function calculateEnvelopeBalances(state, month) {
  const targetIndex = sortedMonths(state).indexOf(month);
  if (targetIndex === -1) {
    return [];
  }

  const balances = new Map();
  const targetRows = [];

  for (const currentMonth of sortedMonths(state).slice(0, targetIndex + 1)) {
    for (const setup of state.monthlySetup.filter((item) => item.month === currentMonth)) {
      const key = envelopeKey(setup.category, setup.subcategory);
      const previous = balances.get(key) || 0;
      const rolloverIn = setup.rollover ? previous : 0;
      const spending = envelopeSpending(state, currentMonth, setup.category, setup.subcategory);
      const available = Number(setup.startingBalance || 0) + rolloverIn + Number(setup.monthlyTarget || 0) - spending;

      balances.set(key, available);

      const row = {
        category: setup.category,
        subcategory: setup.subcategory,
        monthlyTarget: Number(setup.monthlyTarget) || 0,
        startingBalance: Number(setup.startingBalance) || 0,
        rolloverIn,
        spending,
        available,
        overdrawn: available < -MONEY_EPSILON
      };

      if (currentMonth === month) {
        targetRows.push(row);
      }
    }
  }

  return targetRows;
}

export function calculateAccountBalances(state) {
  return state.accounts.map((account) => {
    const accountTransactions = state.transactions.filter((transaction) => transaction.account === account.name);
    const expenses = sum(accountTransactions.filter((item) => item.type === "Expense").map((item) => item.amount));
    const income = sum(accountTransactions.filter((item) => item.type === "Income").map((item) => item.amount));
    const openingBalance = Number(account.openingBalance) || 0;
    const currentBalance = account.type === "Credit Card"
      ? openingBalance - expenses + income
      : openingBalance + income - expenses;

    return {
      account: account.name,
      type: account.type,
      openingBalance,
      currentBalance
    };
  });
}

export function collectWarnings(state, month) {
  const warnings = [];
  const categories = new Set(state.categories.map((item) => item.name));
  const accounts = new Set(state.accounts.map((item) => item.name));
  const subcategories = new Map(state.subcategories.map((item) => [`${item.category}::${item.name}`, item]));
  const subcategoryNames = new Set(state.subcategories.map((item) => item.name));

  for (const envelope of calculateEnvelopeBalances(state, month)) {
    if (envelope.overdrawn) {
      warnings.push({
        severity: "error",
        code: "ENVELOPE_OVERDRAWN",
        message: `${envelope.category} / ${envelope.subcategory} is overdrawn by ${formatMoney(Math.abs(envelope.available))}.`
      });
    }
  }

  if (calculateAvailableToAssign(state, month) < -MONEY_EPSILON) {
    warnings.push({
      severity: "warning",
      code: "OVER_ASSIGNED",
      message: "Monthly envelope funding is greater than income for the selected month."
    });
  }

  for (const transaction of state.transactions) {
    if (!categories.has(transaction.category)) {
      warnings.push({ severity: "warning", code: "MISSING_CATEGORY", message: `Transaction ${transaction.id} uses an unknown category.`, entityId: transaction.id });
    }

    if (!subcategoryNames.has(transaction.subcategory)) {
      warnings.push({ severity: "warning", code: "MISSING_SUBCATEGORY", message: `Transaction ${transaction.id} uses an unknown subcategory.`, entityId: transaction.id });
    } else if (!subcategories.has(`${transaction.category}::${transaction.subcategory}`)) {
      warnings.push({ severity: "warning", code: "SUBCATEGORY_MISMATCH", message: `Transaction ${transaction.id} subcategory does not belong to ${transaction.category}.`, entityId: transaction.id });
    }

    if (!accounts.has(transaction.account)) {
      warnings.push({ severity: "warning", code: "MISSING_ACCOUNT", message: `Transaction ${transaction.id} has no valid account.`, entityId: transaction.id });
    }

    if (!getTransactionMonth(transaction.date)) {
      warnings.push({ severity: "error", code: "INVALID_DATE", message: `Transaction ${transaction.id} has an invalid date.`, entityId: transaction.id });
    }

    if (!Number.isFinite(Number(transaction.amount)) || Number(transaction.amount) <= 0) {
      warnings.push({ severity: "error", code: "INVALID_AMOUNT", message: `Transaction ${transaction.id} has an invalid amount.`, entityId: transaction.id });
    }
  }

  return warnings;
}

function transactionsForMonth(state, month) {
  return state.transactions.filter((transaction) => (transaction.month || getTransactionMonth(transaction.date)) === month);
}

function envelopeSpending(state, month, category, subcategory) {
  return sum(
    transactionsForMonth(state, month)
      .filter((transaction) => transaction.type === "Expense" && transaction.category === category && transaction.subcategory === subcategory)
      .map((transaction) => transaction.amount)
  );
}

function sortedMonths(state) {
  return Array.from(new Set([...MONTHS_2026, ...state.monthlySetup.map((item) => item.month)])).sort();
}

function envelopeKey(category, subcategory) {
  return `${category}::${subcategory}`;
}

function sum(values) {
  return values.reduce((total, value) => total + (Number(value) || 0), 0);
}

function formatMoney(value) {
  return `$${Number(value || 0).toFixed(2)}`;
}

File: backend/app/calculations.py
"""Port of business calculations from src/lib/calculations.js.

Amounts are positive; `type` controls cashflow direction.
"""

from __future__ import annotations

import re
from datetime import date, datetime

from .schema_defaults import MONTHS_2026

MONEY_EPSILON = 0.005


# ---------------------------------------------------------------------------
# Transaction month
# ---------------------------------------------------------------------------

_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


def get_transaction_month(date_string: str) -> str:
    """Return YYYY-MM for a valid YYYY-MM-DD date string, or ''."""
    if not _DATE_RE.match(str(date_string or "")):
        return ""
    parts = date_string.split("-")
    year, month, day = int(parts[0]), int(parts[1]), int(parts[2])
    try:
        d = date(year, month, day)
    except ValueError:
        return ""
    if d.year != year or d.month != month or d.day != day:
        return ""
    return f"{year}-{month:02d}"


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _sum(values):
    return sum(float(v or 0) for v in values)


def _transactions_for_month(transactions: list[dict], month: str) -> list[dict]:
    return [
        t for t in transactions
        if (t.get("month") or get_transaction_month(t.get("date", ""))) == month
    ]


def _sorted_months(state: dict) -> list[str]:
    monthly_setup = state.get("monthlySetup", [])
    months = set(MONTHS_2026)
    for item in monthly_setup:
        months.add(item["month"])
    return sorted(months)


def _envelope_key(category: str, subcategory: str) -> str:
    return f"{category}::{subcategory}"


def _envelope_spending(state: dict, month: str, category: str, subcategory: str) -> float:
    return _sum(
        t["amount"]
        for t in _transactions_for_month(state.get("transactions", []), month)
        if t["type"] == "Expense" and t["category"] == category and t["subcategory"] == subcategory
    )


# ---------------------------------------------------------------------------
# Summary / dashboard
# ---------------------------------------------------------------------------

def summarize_month(state: dict, month: str) -> dict:
    transactions = _transactions_for_month(state.get("transactions", []), month)
    expenses = _sum(t["amount"] for t in transactions if t["type"] == "Expense")
    income = _sum(t["amount"] for t in transactions if t["type"] == "Income")
    category_rows = summarize_categories_for_month(state, month)
    top_category = "None"
    if category_rows:
        top = max(category_rows, key=lambda r: r["spend"])
        top_category = top["category"]

    return {
        "expenses": expenses,
        "income": income,
        "netCashflow": income - expenses,
        "transactionCount": len(transactions),
        "reimbursableTotal": _sum(t["amount"] for t in transactions if t.get("reimbursable")),
        "essentialSpend": _sum(
            t["amount"] for t in transactions
            if t.get("essential") and t["type"] == "Expense"
        ),
        "topCategory": top_category,
    }


def summarize_categories_for_month(state: dict, month: str) -> list[dict]:
    return [
        {
            "category": cat["name"],
            "spend": _sum(
                t["amount"]
                for t in _transactions_for_month(state.get("transactions", []), month)
                if t["type"] == "Expense" and t["category"] == cat["name"]
            ),
            "budget": float(cat.get("defaultBudget", 0) or 0),
        }
        for cat in state.get("categories", [])
    ]


def calculate_available_to_assign(state: dict, month: str) -> float:
    income = _sum(
        t["amount"]
        for t in _transactions_for_month(state.get("transactions", []), month)
        if t["type"] == "Income"
    )
    funding = _sum(
        item["monthlyTarget"]
        for item in state.get("monthlySetup", [])
        if item["month"] == month
    )
    return income - funding


def calculate_envelope_balances(state: dict, month: str) -> list[dict]:
    months = _sorted_months(state)
    try:
        target_index = months.index(month)
    except ValueError:
        return []

    balances: dict[str, float] = {}
    target_rows: list[dict] = []

    for current_month in months[: target_index + 1]:
        for setup in state.get("monthlySetup", []):
            if setup["month"] != current_month:
                continue
            key = _envelope_key(setup["category"], setup["subcategory"])
            previous = balances.get(key, 0.0)
            rollover_in = previous if setup.get("rollover") else 0.0
            spending = _envelope_spending(state, current_month, setup["category"], setup["subcategory"])
            available = float(setup.get("startingBalance", 0) or 0) + rollover_in + float(setup.get("monthlyTarget", 0) or 0) - spending
            balances[key] = available

            row = {
                "category": setup["category"],
                "subcategory": setup["subcategory"],
                "monthlyTarget": float(setup.get("monthlyTarget", 0) or 0),
                "startingBalance": float(setup.get("startingBalance", 0) or 0),
                "rolloverIn": rollover_in,
                "spending": spending,
                "available": available,
                "overdrawn": available < -MONEY_EPSILON,
            }
            if current_month == month:
                target_rows.append(row)

    return target_rows


def calculate_account_balances(state: dict) -> list[dict]:
    result = []
    for account in state.get("accounts", []):
        account_transactions = [
            t for t in state.get("transactions", [])
            if t.get("account") == account["name"]
        ]
        expenses = _sum(t["amount"] for t in account_transactions if t["type"] == "Expense")
        income = _sum(t["amount"] for t in account_transactions if t["type"] == "Income")
        opening_balance = float(account.get("openingBalance", 0) or 0)
        if account.get("type") == "Credit Card":
            current_balance = opening_balance - expenses + income
        else:
            current_balance = opening_balance + income - expenses

        result.append({
            "account": account["name"],
            "type": account.get("type", "Bank"),
            "openingBalance": opening_balance,
            "currentBalance": current_balance,
        })
    return result


def collect_warnings(state: dict, month: str) -> list[dict]:
    """Return list of WarningItem dicts."""
    warnings: list[dict] = []

    category_names = {c["name"] for c in state.get("categories", [])}
    account_names = {a["name"] for a in state.get("accounts", [])}
    subcategory_map = {f"{s['category']}::{s['name']}": s for s in state.get("subcategories", [])}
    subcategory_names = {s["name"] for s in state.get("subcategories", [])}

    # overdrawn envelopes
    for envelope in calculate_envelope_balances(state, month):
        if envelope["overdrawn"]:
            warnings.append({
                "severity": "error",
                "code": "ENVELOPE_OVERDRAWN",
                "message": f"{envelope['category']} / {envelope['subcategory']} is overdrawn by ${abs(envelope['available']):.2f}.",
                "entityId": None,
            })

    # over-assigned
    if calculate_available_to_assign(state, month) < -MONEY_EPSILON:
        warnings.append({
            "severity": "warning",
            "code": "OVER_ASSIGNED",
            "message": "Monthly envelope funding is greater than income for the selected month.",
            "entityId": None,
        })

    # per-transaction
    for t in state.get("transactions", []):
        tid = t.get("id", "")

        if t["category"] not in category_names:
            warnings.append({
                "severity": "warning", "code": "MISSING_CATEGORY",
                "message": f"Transaction {tid} uses an unknown category.",
                "entityId": tid,
            })
        if t["subcategory"] not in subcategory_names:
            warnings.append({
                "severity": "warning", "code": "MISSING_SUBCATEGORY",
                "message": f"Transaction {tid} uses an unknown subcategory.",
                "entityId": tid,
            })
        elif f"{t['category']}::{t['subcategory']}" not in subcategory_map:
            warnings.append({
                "severity": "warning", "code": "SUBCATEGORY_MISMATCH",
                "message": f"Transaction {tid} subcategory does not belong to {t['category']}.",
                "entityId": tid,
            })
        if t["account"] not in account_names:
            warnings.append({
                "severity": "warning", "code": "MISSING_ACCOUNT",
                "message": f"Transaction {tid} has no valid account.",
                "entityId": tid,
            })
        if not get_transaction_month(t.get("date", "")):
            warnings.append({
                "severity": "error", "code": "INVALID_DATE",
                "message": f"Transaction {tid} has an invalid date.",
                "entityId": tid,
            })
        amt = t.get("amount", 0)
        if not isinstance(amt, (int, float)) or amt != amt or amt <= 0:
            warnings.append({
                "severity": "error", "code": "INVALID_AMOUNT",
                "message": f"Transaction {tid} has an invalid amount.",
                "entityId": tid,
            })

    return warnings


def build_derived_state(state: dict) -> dict:
    """Return a DerivedState-shaped dict for the selected month."""
    month = state.get("selectedMonth", "2026-01")
    return {
        "summary": summarize_month(state, month),
        "categoryRows": summarize_categories_for_month(state, month),
        "availableToAssign": calculate_available_to_assign(state, month),
        "envelopeRows": calculate_envelope_balances(state, month),
        "accountBalances": calculate_account_balances(state),
        "warnings": collect_warnings(state, month),
    }
